### PR TITLE
many: add snapcraftctl command for scriptlets

### LINKED ISSUE #1675: Base logic for metadata setters set-
Estimate: 4 days

We currently have three scriptlets that cover the build step only: prepare, build, and install. We want scriptlets around all lifecycle steps: pull, build, stage, and prime. Note that introducing an install step is not part of this roadmap.

There are some complications here that I want to point out. The stage and prime steps are completely handled by the core right now. As soon as we change that and provide the ability to override them, we must account for the possibility that files in stage will be changed from what they were in install, etc. Which means the stage dir must flow to the prime dir, which it does not do today (the file list is retrieved from stage, but the files themselves come from the parts’ install dirs).

Note that the stanzas for pull and build can be completed without such a redesign.

### PR DESCRIPTION
**Pull request by kyrofa**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh unit`?

-----

There is the desire for scriptlets to be able to communicate with Snapcraft, e.g. to request that it use a specific name, or version, or do a specific action, such as fetch stage packages, or run a lifecycle step.

This PR resolves #1675 by adding the basic framework for this communication, adding a new CLI
utility called `snapcraftctl` that is used to communicate with the parent Snapcraft process. It proves this out by adding a new subcommand to this utility called `build` which, if called, will run the plugin's build step. It also adds the `overide-build` scriptlet, which can use `snapcraftctl build` like so:

    # ...
    override-build: |
      snapcraftctl build
    # ...

Note that, as a side-effect of this work, LP: [#1751749](https://bugs.launchpad.net/snapcraft/+bug/1751749) is fixed as well.

### CODE REVIEW

File: snapcraft/internal/errors.py [hunk 1]
```diff
@@ -517,3 +517,17 @@ class InvalidExtractorValueError(MetadataExtractionError):
 
     def __init__(self, path: str, extractor_name: str) -> None:
         super().__init__(path=path, extractor_name=extractor_name)
+
+
+class ScriptletError(SnapcraftError):
+    pass
```
Comment (sergiusens) [inline snapcraft/internal/errors.py]: I've noticed that instead of pass we could also add a simple docstring to this explaining that it is a base class for others (and even if it will bring in inconcistency, we should add Base to the name, wdyt?)

File: snapcraft/internal/pluginhandler/_runner.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+
+class Runner:
+    """The Runner class is responsible for orchestrating scriptlets."""
+
+    # FIXME: Need to quote builtin_functions typing because of
+    # https://github.com/python/typing/issues/259 which is fixed in Python
+    # 3.5.3.
+    def __init__(self, *, part_properties: Dict[str, Any], builddir: str,
+                 builtin_functions: 'Dict[str, Callable[..., None]]') -> None:
```
Comment (sergiusens) [inline snapcraft/internal/pluginhandler/_runner.py]: try creating a type from `'Dict[str, Callable[..., None]]`, look at elf.py for inspiration

Comment (kyrofa) [reply snapcraft/internal/pluginhandler/_runner.py]: I can't actually get it to work with the quoting, keeps telling me it's an invalid type.

File: snapcraft/internal/errors.py (hunk 1 above)
Comment (kyrofa) [reply snapcraft/internal/errors.py]: Agreed on both points, done.

File: snapcraft/internal/pluginhandler/_runner.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+                ['/bin/sh', '-e', '-c', script], cwd=self._builddir)
+
+            status = None
+            try:
+                while status is None:
+                    function_call = call_fifo.read()
+                    if function_call:
+                        self._handle_builtin_function(
+                            scriptlet_name, function_call.strip())
+                        feedback_fifo.write('success\n')
```
Comment (sergiusens) [inline snapcraft/internal/pluginhandler/_runner.py]: why do we write `success\n`?

File: snapcraft/internal/pluginhandler/_runner.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+                    status = process.poll()
+            finally:
+                call_fifo.close()
+                feedback_fifo.close()
+
+            if status:
+                raise errors.ScriptletRunError(
+                    scriptlet_name=scriptlet_name, code=status)
```
Comment (sergiusens) [inline snapcraft/internal/pluginhandler/_runner.py]: this I think solves a bug with scriptlet errors you might want to take attributes for

File: snapcraft/internal/pluginhandler/_runner.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+    def _handle_builtin_function(self, scriptlet_name, function_call):
+        function_call_components = shlex.split(function_call)
+        if len(function_call_components) == 0:
+            # This means a snapcraft developer screwed up. Should never be
```
Comment (sergiusens) [inline snapcraft/internal/pluginhandler/_runner.py]: let's try and avoid the word "screwed", it feels a bit confrontational :-)

File: snapcraft/internal/pluginhandler/_runner.py (hunk 3 above)
Comment (kyrofa) [reply snapcraft/internal/pluginhandler/_runner.py]: All that's actually required is a '\n' to get the [read](https://github.com/snapcore/snapcraft/pull/2002/files#diff-8b247ab4b649deb9ee488e09837eba39R33) to stop waiting for the call to be handled, but that felt a little too magical for my taste :stuck_out_tongue: . This could eventually grow into actually passing information back to the scriptlet from the builtins if we want.

File: snapcraft/internal/pluginhandler/_runner.py (hunk 4 above)
Comment (kyrofa) [reply snapcraft/internal/pluginhandler/_runner.py]: This in combination with the fact that we're no longer using a temporary script should indeed fix LP: [#1751749](https://bugs.launchpad.net/snapcraft/+bug/1751749), thank you for reminding me.

File: snapcraft/internal/pluginhandler/_runner.py (hunk 5 above)
Comment (kyrofa) [reply snapcraft/internal/pluginhandler/_runner.py]: Of course, good call. Fixed.

File: snapcraft/internal/pluginhandler/_runner.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+            status = None
+            try:
+                while status is None:
+                    function_call = call_fifo.read()
+                    if function_call:
+                        self._handle_builtin_function(
+                            scriptlet_name, function_call.strip())
+                        feedback_fifo.write('success\n')
+                    status = process.poll()
```
Comment (chipaca) [inline snapcraft/internal/pluginhandler/_runner.py]: I'd add a time.sleep(.1) here (or .01 if you're impatient :-)

File: snapcraft/internal/pluginhandler/_runner.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+                '{!r} scriptlet called an undefined builtin function: '
+                '{}'.format(scriptlet_name, function_call)) from e
+
+
+class _NonBlockingFifo:
+
+    def __init__(self, *, path: str, permissions) -> None:
+        os.mkfifo(path)
+        self.path = path
+        self._permissions = permissions
```
Comment (chipaca) [inline snapcraft/internal/pluginhandler/_runner.py]: I'd get rid of 'permissions' entirely, and always open the fifo `O_RDWR`. This'd allow you to drop the `_fifo_open` method entirely and just open in in `__init__`.

Comment (kyrofa) [reply snapcraft/internal/pluginhandler/_runner.py]: Ah! A genius you are! That was the part of this that was really bothering me.

Comment (chipaca) [reply snapcraft/internal/pluginhandler/_runner.py]: merely a reader of manpages :-D

File: snapcraft/internal/pluginhandler/_runner.py (hunk 6 above)
Comment (kyrofa) [reply snapcraft/internal/pluginhandler/_runner.py]: Yeah good call, thank you :) .

File: snapcraft/cli/__main__.py [hunk 8]
```diff
@@ -21,6 +21,7 @@
 import subprocess
 
 from ._runner import run
+from .snapcraftctl._runner import run as run_snapcraftctl  # noqa
```
Comment (sergiusens) [inline snapcraft/cli/__main__.py]: why is this needed here?

Comment (sergiusens) [reply snapcraft/cli/__main__.py]: oh, never mind, console_scripts

File: snapcraft/cli/snapcraftctl/_runner.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+import click
+
+from snapcraft.internal.errors import SnapcraftEnvironmentError
+from snapcraft.cli._errors import exception_handler
+from snapcraft.internal import log
+
+
+@click.group()
+@click.option('--debug', '-d', is_flag=True)
```
Comment (sergiusens) [inline snapcraft/cli/snapcraftctl/_runner.py]: it would be nice if this could be picked up from the environment as well `SNAPCRAFT_DEBUG` (click supports this) so then it is hooked up with snapcraft's flag too

File: snapcraft/cli/__main__.py (hunk 8 above)
Comment (kyrofa) [reply snapcraft/cli/__main__.py]: Yeah, didn't want to duplicate the whole "set UTF-8" thing.

File: snapcraft/cli/snapcraftctl/_runner.py (hunk 9 above)
Comment (kyrofa) [reply snapcraft/cli/snapcraftctl/_runner.py]: Oh good call, fixed in the most recent commit.

File: bin/snapcraftctl [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+snapcraft.cli.__main__.run_snapcraftctl(prog_name=\"snapcraftctl\")
+' ""$@"
+
+
+# We don't actually want a 100% clean environment. Pass on the SNAP variables,
+# locale settings, and environment variables required by snapcraftctl itself.
+/usr/bin/env -i -- sh -<<END
+# Required for snapcraftctl to actually find snapcraft when snapped via
+# sitecustomize
+export SNAP="$SNAP"
```
Comment (kyrofa) [inline bin/snapcraftctl]: Only set if defined.
